website_blog_layout_options/models/blog_post.py

```diff
@@ -0,0 +1,100 @@
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp import models, fields, api
+from lxml import etree
+
+
+class BlogPost(models.Model):
+
+    _inherit = 'blog.post'
+
+    @api.one
+    @api.onchange('extract_auto', 'display_type', 'content', 'teaser_input')
+    def _extract_teaser(self):
+        if self.display_type == "teaser":
+            # no empty teasers
+            if (self.teaser_input or '') and (not self.extract_auto):
+                self.teaser = self.teaser_input
+            else:
+                res = ""
+                # limit length to roughly 3-4 lines.
+                teaser_length = 500
+                parser = etree.HTMLParser()
+                if self.content:
+                    tree = etree.fromstring(self.content, parser)
+                    paragraphs = tree.xpath('//p')
+                    # get the first non empty paragraph
+                    for paragraph in paragraphs:
+                        if paragraph.text and len(res) < teaser_length:
+                            res = res + paragraph.text + '\n'
+                        else:
+                            break
+                    # trim it to the intended length
+                    self.teaser = res[:teaser_length] + " ..."
+                    self.teaser_input = res[:teaser_length] + " ..."
+                else:
+                    # has no teaser or content,  just revert.
+                    # frontend controls needed not to have a bad workflow.
+                    # content cannot be inserted in backend by default.
+                    # add content to backend in view
+                    self.display_type = "no_teaser"
+
+    @api.onchange('blog_id')
+    def set_new_default(self):
+        self.background_image_show = self.blog_id.background_image_show
+
+    background_image_show = fields.Selection(
+        string="Type of header image on blog post",
+        selection=lambda self: self.blog_id._get_image_options(),
+        default=lambda self: self.blog_id.background_image_show,
+        required=True,
+        help="Choose if how you want to display the blog post:"
+        "Just the title above the post, a small header image"
+        "above the blog post title, or a big full screen image,"
+        "before showing the post, (odoo default)")
+
+    thumbnail = fields.Many2one(
+        string='Blog Post Thumbnail',
+        comodel_name='ir.attachment',
+        help='A small image shown in teaser and content'
```

Comment:
add a domain for image file formats. but why not a function field that uploads an image to the attachments or a plain binary field?